scripts: Convert gen_offset_header to Python

lpereira opened the pull request:
By having this as a Python script rather than a host executable, this
should simplify the build process on non-Unix platforms.  The
'pyelftools' library is required for this, and can be installed with:

	pip install pyelftools

Jira: ZEP-2062
Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

tyrantlizard77 reviewed (changes requested):
This PR must include changes to the first-time setup doucmentation for all 3 host arches (mac/linux/windows) that pyelftools must be installed.

We also need to install it on the CI machines for this PR to pass CI

Python code looks fine.

mbolivar commented:
This is also 'python3-pyelftools' on Debian based distros, FYI. (Some admins may prefer using the package manager to pip).

pfalcon commented on scripts/gen_offset_header.py:
```diff
@@ -0,0 +1,56 @@
+import sys
+
+def gen_offset_header(input_name, input_file, output_file):
+  obj = ELFFile(input_file)
```
Indentation in this file doesn't conform to PEP8.

andrewboie replied:
agreed I would prefer to see 4-space indentation

pfalcon reviewed (commented):
I'm personally all for replacing obfuscated C sources with Python ones, but I'm not sure if this change will make it better for everyone - this adds dependency on 3rd-party Python module, whereas original tool didn't have additional dependencies.

andrewboie commented on scripts/gen_offset_header.py:
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/python
```
we have standardized on Python 3 for all Zephyr python code, please convert and change this line to 
#!/usr/bin/env python3

andrewboie commented:
CI does not like the hard tab in the commit message, BTW

andrewboie commented:
@pfalcon we already require several pip modules to be installed and we are converting all C-based build tools to Python to make Windows build support simpler

pfalcon commented:
I see, cool.

andrewboie commented on scripts/gen_offset_header.py:
```diff
@@ -0,0 +1,56 @@
+  obj = ELFFile(input_file)
+
+  symtab = obj.get_section_by_name('.symtab')
+  if not symtab:
```
I suggest throwing an exception in this and the below isinstance check if they fail

andrewboie commented:
I'm getting a lot of build errors, seems like there is a stray space being inserted?

                 from /projects/zephyr2/arch/arc/core/swap.S:18:
/projects/zephyr2/sanity-out/arduino_101_sss/tests/kernel/mem_slab/mslab/test/include/generated/offsets.h:8:26: warning: extra tokens at end of #ifndef directive
 #ifndef __INCLUDE_GUARD_ 5d8d48e39d92163__
                          ^~~~~~~~~~~~~~~~~

andrewboie commented on scripts/gen_offset_header.py:
```diff
@@ -0,0 +1,62 @@
+        print(".symtab section is not an a symbol table")
+        return 1
+
+    include_guard = "__INCLUDE_GUARD_%16x__" % abs(input_name.__hash__())
```
i think you need %016x to zero-lead the value otherwise there are gonna be spaces

lpereira commented:
Pull request to install pyelftools on Shippable is in the `ci-dockerfiles` repository, here: https://github.com/zephyrproject-rtos/ci-dockerfiles/pull/1

dbkinder reviewed (dismissed):
+1 for doc changes

andrewboie commented on doc/getting_started/installation_linux.rst:
```diff
@@ -64,7 +65,7 @@ Install the required packages in a Fedora host system with:
    $ sudo dnf install git make gcc glibc-static \
 	 libstdc++-static python3-ply ncurses-devel \
-	 python-yaml dfu-util dtc
+	 python-yaml dfu-util dtc python3-pyelftools
```
this installs version 0.22 which is too old.
Add to pip line instead.

andrewboie commented on doc/getting_started/installation_linux.rst:
```diff
@@ -55,7 +55,8 @@ Install the required packages in a Ubuntu host system with:
    $ sudo apt-get install git make gcc g++ python3-ply ncurses-dev \
-	 python3-yaml dfu-util device-tree-compiler
+	 python3-yaml dfu-util device-tree-compiler \
+	 python3-pyelftools
```
i don't know what version this installs.
if it's not 0.24, recommend install via pip instead.

lpereira commented:
@andrewboie Updated docs to suggest pyelftools installation through pip.

andrewboie commented:
LGTM.
I really hope we don't have to wait until next week to get pyelftools on the CI machines....

galak commented:
@andrewboie I'll merge the docker update, but would like it to pass CI.

andrewboie commented:
@galak agreed.
@nashif  said he would spend some time looking at this today.
I think the docker patch may need to be updated, unless we are using Ubuntu 16.10 or later we will need to install with pip, the library version must be 0.24 or later.

andrewboie commented:
If you rebase I think we can merge this

nashif commented:
> I'm personally all for replacing obfuscated C sources with Python ones, but I'm not sure if this change will make it better for everyone - this adds dependency on 3rd-party Python module, whereas original tool didn't have additional dependencies.

The original tool had dependency on a host compiler, and this is a major dependency compared to a python module. We do need python anyways and it is easier to get install compared with a host compiler, for example on windows...

nashif commented on Kbuild:
```diff
@@ -51,15 +51,15 @@ arch/$(ARCH)/core/offsets/offsets.o: arch/$(ARCH)/core/offsets/offsets.c $(KCONF
+	$(Q)set -e;                                                          \
+	$(kecho) '  CHK     $@';                                             \
+	mkdir -p $(dir $@);                                                  \
+	$(PYTHON) $(srctree)/scripts/gen_offset_header.py -i $(1) -o $@.tmp; \
```
this will call the script with python which by default on many systems will be python2. Call the script directly instead.

nashif commented on scripts/gen_offset_header.py:
```diff
@@ -0,0 +1,62 @@
+    args = parser.parse_args()
+
+    input_file = open(args.input, 'rb')
+    output_file = open(args.output, 'wb')
```
why are we writing binary? This fails with python3

nashif commented on scripts/gen_offset_header.py:
```diff
@@ -0,0 +1,62 @@
+def gen_offset_header(input_name, input_file, output_file):
+    obj = ELFFile(input_file)
+
+    symtab = obj.get_section_by_name(b'.symtab')
```
This is not working for me, symtab does not exist, I quickly hacked this to 

~~~~
-    symtab = obj.get_section_by_name(b'.symtab')
-    if not symtab:
-        print("Could not find .symtab section in %s" % input_name)
-        return 1
-
-    if not isinstance(symtab, SymbolTableSection):
-        print(".symtab section is not an a symbol table")
-        return 1
+    for section in obj.iter_sections():
+        if isinstance(section, SymbolTableSection):
+            symtab = section
~~~~

and now it works.

andrewboie commented:
hmm I seem to remember documentation changes in one version of this patch set, but they seem to not  be in this latest version?

lpereira commented:
Yes, that's strange; the documentation was updated in a previous version. Sending a new one in a few minutes.

nashif commented:
> Yes, that's strange; the documentation was updated in a previous version. Sending a new one in a few minutes.

Wait, this was modified globally in another patch already and was conflicting. I did the change.

No need to document individual python modules for every OS now, it is all listed in the requirements.txt file in one file now.

lpereira commented:
The documentation changes are not required any longer, since all Python dependencies are descibed in `scripts/requirements.txt`.

lpereira committed:
scripts: Convert gen_offset_header to Python

By having this as a Python script rather than a host executable, this
should simplify the build process on non-Unix platforms.

With this change, pyelftools is now required to build Zephyr.  Please
consult the getting started documentation for your host platform for
installation instructions.

Jira: ZEP-2062

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira commented:
Rebased and addressed @nashif's comments.
